File: f77code/Chp08/rk45.f
C
C PAGE 326: NUMERICAL MATHEMATICS AND COMPUTING, CHENEY/KINCAID, 1985
C
C FILE: RK45.FOR
C
C RUNGE-KUTTA-FEHLBERG METHOD FOR SOLVING AN INITIAL VALUE PROBLEM (RK45,F)  
C
      SUBROUTINE RK45(F,T,X,H,EST)    
      DATA  C21,C31,C32, C41,C42,C43, C51,C52,C53,C54,    
     A      C61,C62,C63,C64,C65, A1,A3,A4,A5, B1,B3,B4,B5, B6, C40  
     B  /0.25,0.09375,0.28125,
     C  0.87938097405553,-3.2771961766045,3.3208921256258,
     D  2.0324074074074,-8.0,7.1734892787524,-0.20589668615984,     
     E  -0.2962962962963,2.0,-1.3816764132554,0.45297270955166,-0.275,
     F  0.11574074074074,0.54892787524366,0.5353313840156,-0.2,     
     G  0.11851851851852,0.51898635477583,0.50613149034201,-0.18,   
     H  0.036363636363636, 0.92307692307692/    
      F1 = H*F(T,X) 
      F2 = H*F(T+ 0.25*H,X + C21*F1)  
      F3 = H*F(T+0.375*H,X + C31*F1 + C32*F2)   
      F4 = H*F(T+C40*H,X + C41*F1 + C42*F2 + C43*F3)  
      F5 = H*F(T+H    ,X + C51*F1 + C52*F2 + C53*F3 + C54*F4)       
      F6 = H*F(T+0.5*H,X + C61*F1 + C62*F2 + C63*F3 + C64*F4 + C65*F5)
      X5 = X + B1*F1 + B3*F3 + B4*F4 + B5*F5 + B6*F6      
      X  = X + A1*F1 + A3*F3 + A4*F4 + A5*F5    
      T = T + H   
      EST = ABS(X - X5)     
      RETURN      
      END 
  
      EXTERNAL F
      DATA T/1.0/, X/2.0/, H/7.8125E-3/, NSTEP/72/
      PRINT *,T,X 
      DO 2 I=1,NSTEP
      CALL RK45(F,T,X,H,EST)
      PRINT *,T,X,EST 
   2  CONTINUE
      STOP
      END 
  
      FUNCTION F(T,X) 
      F = 2.0+(X-T-1.0)**2  
      RETURN
      END 

File: c++code/Chp08/rk45.c
/****************************************************************/
/* Module : rk45.c 						*/
/* Section: 8.3							*/
/* Cheney-Kincaid, Numerical Mathematics and Computing, 4th ed, */
/* Brooks/Cole Publ. Co.                                        */
/* ISBN 0-534-20112-1                                           */
/* Copyright (c) 1999.  All rights reserved.                    */
/* For educational use with the Cheney-Kincaid textbook.        */
/* Absolutely no warranty implied or expressed.                 */   
/* 								*/
/* Programmer: Hidajaty Thajeb					*/
/* Description:	Runge-Kutta-Fehlberg method for solving an 	*/
/*              initial value problem.                          */
/* Comment:                                                     */
/*                                                              */
/****************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

inline float f(float t, float x) { return 2.0 + (x - t - 1.0)*(x - t - 1.0); }
void rk45(float (float, float), float&, float&, float, float&);

int main()
{
  const int n = 72;
  float a = 1.0, b = 1.5625, x = 2.0;
  int k; 
  float t, h, epsilon;

  h = (b - a)/n;
  t = a;
  printf("0 x(%1.3f) = %1.5f\n", t, x);

  for (k = 1; k <= n;  k++)
  {
    rk45(f, t, x, h, epsilon);
    printf("k = %d: x(%1.5f) = %1.5f, e = %f\n", k, t, x, epsilon);
  }
  return 0;
}

//--------------------------------------------------------------------------
void rk45(float f(float, float), float& t, float& x, float h, float& epsilon)
{

  float c20 = 0.25, c21 = 0.25;
  float c30 = 0.375, c31 = 0.09375, c32 = 0.28125;
  float c40,c41, c42,c43;
  float c51, c52 = -8.0, c53, c54;
  float c60 = 0.5, c61, c62 = 2, c63, c64;
  float c65 = -0.275;
  float a1, a2 = 0, a3, a4, a5 = -0.2;
  float b1, b2 = 0, b3, b4, b5= -0.18, b6;
  float K1, K2, K3, K4, K5, K6, x4;

  c40 = (float) 12/ (float) 13;
  c41 = (float) 1932/(float) 2197;
  c42 = (float) -7200/(float) 2197;
  c43 = (float) 7296/(float) 2197;
  c51 = c53 = (float) 439/ (float) 216;
  c54 = (float) -845/(float) 4104;
  c61 = (float) -8/(float) 27;
  c63 = (float) -3544/(float) 2565;
  c64 = (float) 1859/(float) 4104;
  a1 = (float) 25/(float) 216;
  a3 = (float) 1408/(float) 2565;
  a4 = (float) 2197/(float) 4104;
  b1 = (float) 16/(float) 135;
  b3 = (float) 6656/(float) 12825;
  b4 = (float) 28561/(float) 56430;
  b6 = (float) 2/(float) 55;


  K1 = h*f(t, x);
  K2 = h*f(t + c20*h, x + c21*K1);
  K3 = h*f(t + c30*h, x + c31*K1 + c32*K2);
  K4 = h*f(t + c40*h, x + c41*K1 + c42*K2 + c43*K3);
  K5 = h*f(t + h, x + c51*K1 + c52*K2 + c53*K3 + c54*K4 );
  K6 = h*f(t + c60*h, x + c61*K1 + c62*K2 + c63*K3 + c64*K4 + c65*K5);

  x4 = x + a1*K1 + a3*K3 + a4*K4 + a5*K5;
  x += b1*K1 + b3*K3 + b4*K4 + b5*K5 + b6*K6;
  t += h;
  epsilon = fabs(x - x4);

}






